File: testdata/fixtures/java/CWE-295.java
package fixture;

// CWE-295: Improper Certificate Validation - TrustManager accepts all certs

import java.security.cert.X509Certificate;
import javax.net.ssl.*;

class Cwe295TrustAll {

    void vulnerable1() throws Exception {
        TrustManager[] trustAll = new TrustManager[] {
            new X509TrustManager() {
                // ruleid: java-tls-insecure
                public void checkClientTrusted(X509Certificate[] chain, String authType) {}
                // ruleid: java-tls-insecure
                public void checkServerTrusted(X509Certificate[] chain, String authType) {}
                public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
            }
        };
        SSLContext.getInstance("TLS").init(null, trustAll, new java.security.SecureRandom());
    }

    void safe() throws Exception {
        // ok: java-tls-insecure
        TrustManager[] trust = new TrustManager[] {
            new X509TrustManager() {
                public void checkClientTrusted(X509Certificate[] chain, String authType) throws java.security.cert.CertificateException {
                    if (chain == null || chain.length == 0) throw new java.security.cert.CertificateException("empty");
                }
                public void checkServerTrusted(X509Certificate[] chain, String authType) throws java.security.cert.CertificateException {
                    if (chain == null || chain.length == 0) throw new java.security.cert.CertificateException("empty");
                }
                public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
            }
        };
    }

    void noTrustManager() {
        // ok: java-tls-insecure
        System.out.println("no TLS here");
    }
}
    void vulnerable2() throws Exception {
        TrustManager[] trustAll2 = new TrustManager[] {
            new X509TrustManager() {
                // ruleid: java-tls-insecure
                public void checkClientTrusted(X509Certificate[] chain, String authType) {}
                // ruleid: java-tls-insecure
                public void checkServerTrusted(X509Certificate[] chain, String authType) {}
                public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
            }
        };
        SSLContext.getInstance("TLS").init(null, trustAll2, new java.security.SecureRandom());
    }

    void safeWithCheck() throws Exception {
        // ok: java-tls-insecure
        TrustManager[] tm = new TrustManager[] {
            new X509TrustManager() {
                public void checkClientTrusted(X509Certificate[] chain, String authType) throws java.security.cert.CertificateException {
                    if (chain == null) throw new java.security.cert.CertificateException();
                }
                public void checkServerTrusted(X509Certificate[] chain, String authType) throws java.security.cert.CertificateException {
                    if (chain == null) throw new java.security.cert.CertificateException();
                }
                public X509Certificate[] getAcceptedIssuers() { return null; }
            }
        };
    }

File: testdata/fixtures/scala/CWE-295.scala
// CWE-295: Improper Certificate Validation - disabled verification
//
// TrustManager that does nothing or SSLContext initialized with null trust
// disables certificate verification.

import javax.net.ssl.{SSLContext, X509TrustManager}
import java.security.cert.X509Certificate

object Cwe295Tls {
  def insecure1(): Unit = {
    val tm = new X509TrustManager {
      def getAcceptedIssuers(): Array[X509Certificate] = Array.empty
      def checkClientTrusted(chain: Array[X509Certificate], authType: String): Unit = {}
      def checkServerTrusted(chain: Array[X509Certificate], authType: String): Unit = {}
    }
    // ruleid: scala-tls-insecure
    SSLContext.getInstance("TLS").init(null, Array(tm), null)
  }

  def insecure2(): Unit = {
    // ruleid: scala-tls-insecure
    SSLContext.getInstance("SSL").init(null, null, null)
  }

  def insecure3(): Unit = {
    // ruleid: scala-tls-insecure
    SSLContext.getInstance("TLSv1").init(null, Array.empty, null)
  }

  def safe(): Unit = {
    // ok: scala-tls-insecure
    val ctx = SSLContext.getInstance("TLSv1.2")
    ctx.init(null, null, new java.security.SecureRandom())
    // ok: scala-tls-insecure
    SSLContext.getInstance("TLSv1.3")
    // ok: scala-tls-insecure
    val safeTm = new X509TrustManager {
      def getAcceptedIssuers(): Array[X509Certificate] = Array.empty
      def checkClientTrusted(chain: Array[X509Certificate], authType: String): Unit = throw new Exception()
      def checkServerTrusted(chain: Array[X509Certificate], authType: String): Unit = throw new Exception()
    }
  }
}
